streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/ClientTagAwareStandbyTaskAssignor.java

```diff
@@ -0,0 +1,203 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.streams.processor.internals.assignment;
+
+import org.apache.kafka.streams.processor.TaskId;
+import org.apache.kafka.streams.processor.internals.assignment.AssignorConfiguration.AssignmentConfigs;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.Set;
+import java.util.UUID;
+
+import static org.apache.kafka.streams.processor.internals.assignment.StandbyTaskAssignmentUtils.computeTasksToRemainingStandbys;
+
+/**
+ * Distributes standby tasks over different tag dimensions.
+ * Only tags specified via {@link AssignmentConfigs#rackAwareAssignmentTags} are taken into account.
+ * Standby task distribution is on a best-effort basis. For example, if there are not enough clients available
+ * on different tag dimensions compared to an active and corresponding standby task,
+ * in that case, the algorithm will fall back to distributing tasks on least-loaded clients.
+ */
+class ClientTagAwareStandbyTaskAssignor implements StandbyTaskAssignor {
+    private static final Logger log = LoggerFactory.getLogger(ClientTagAwareStandbyTaskAssignor.class);
+
+    @Override
+    public boolean assign(final Map<UUID, ClientState> clients,
+                          final Set<TaskId> allTaskIds,
+                          final Set<TaskId> statefulTaskIds,
+                          final AssignorConfiguration.AssignmentConfigs configs) {
+        final int numStandbyReplicas = configs.numStandbyReplicas;
+        final Set<String> rackAwareAssignmentTags = new HashSet<>(configs.rackAwareAssignmentTags);
+        final Map<TaskId, UUID> statefulTasksWithClients = new HashMap<>();
+
+        statefulTaskIds.forEach(statefulTaskId -> clients.forEach((uuid, clientState) -> {
+            if (clientState.activeTasks().contains(statefulTaskId)) {
+                statefulTasksWithClients.put(statefulTaskId, uuid);
+            }
+        }));
+
+        final Map<TaskId, Integer> tasksToRemainingStandbys = computeTasksToRemainingStandbys(
+            numStandbyReplicas,
+            allTaskIds
+        );
+
+        final Map<String, Set<String>> tagKeyToTagValues = new HashMap<>();
+        final Map<String, Set<UUID>> tagValueToClients = new HashMap<>();
+
+        fillClientsTagStatistics(clients, tagValueToClients, tagKeyToTagValues);
+
+        statefulTasksWithClients.forEach((taskId, clientId) -> assignStandbyTasksForActiveTask(
+            numStandbyReplicas,
+            taskId,
+            clientId,
+            rackAwareAssignmentTags,
+            clients,
+            tasksToRemainingStandbys,
+            tagKeyToTagValues,
+            tagValueToClients
+        ));
```

Comment:
Map `statefulTasksWithClients` is only used to iterate over its entries. I think it would be better to use the following nested loops and remove `statefulTasksWithClients`:

```suggestion
        for (final TaskId statefulTaskId : statefulTaskIds) {
            for (final Map.Entry<UUID, ClientState> entry : clients.entrySet()) {
                final UUID clientId = entry.getKey();
                final ClientState clientState = entry.getValue();

                if (clientState.activeTasks().contains(statefulTaskId)) {
                    assignStandbyTasksForActiveTask(
                        numStandbyReplicas,
                        statefulTaskId,
                        clientId,
                        rackAwareAssignmentTags,
                        clients,
                        tasksToRemainingStandbys,
                        tagKeyToTagValues,
                        tagValueToClients
                    );
                }
            }
        }
```